### PR TITLE
Throttling

### PR DESCRIPTION
**Pull request by fosskers**

### CODE REVIEW

File: chainweb.cabal [hunk 1]
```diff
@@ -1,7 +1,7 @@
 cabal-version: 2.2
 
 name:         chainweb
-version:      1.0.2
+version:      1.1.0
```
Comment (fosskers) [inline chainweb.cabal]: This PR makes breaking changes to config.

File: src/Chainweb/Chainweb.hs [hunk 2]
```diff
@@ -674,18 +697,22 @@ withChainwebInternal conf logger peer rocksDb dbDir nodeid resetDb inner = do
...
 
-mkPutPeerThrottler :: Integral a => a -> IO (Throttle Address)
+mkPutPeerThrottler :: Double -> IO (Throttle Address)
 mkPutPeerThrottler rate = mkThrottler 5 rate $ \r ->
     elem "peer" (pathInfo r) && requestMethod r == "PUT"
 
+mkLocalThrottler :: Double -> IO (Throttle Address)
+mkLocalThrottler rate = mkThrottler 5 rate (checkPathPrefix path)
+  where
+    path = ["pact", "api", "v1", "local"]
```
Comment (fosskers) [inline src/Chainweb/Chainweb.hs]: This is the correct path?

File: src/Chainweb/Chainweb.hs [hunk 3]
```diff
@@ -697,18 +724,17 @@ checkPathPrefix endpoint r = endpoint `isPrefixOf` drop 3 (pathInfo r)
...
     -> (Request -> Bool)
         -- ^ Predicate to select requests that are throttled
     -> IO (Throttle Address)
-mkThrottler e rate c = initThrottler (defaultThrottleSettings $ TimeSpec (int e) 0) -- expiration
-    { throttleSettingsRate = int rate -- number of allowed requests per period
-    , throttleSettingsPeriod = 1 / micro -- 1 second (measured in usec)
-    , throttleSettingsBurst = 2 * int rate
+mkThrottler e rate c = initThrottler (defaultThrottleSettings $ TimeSpec (ceiling e) 0) -- expiration
+    { throttleSettingsRate = rate -- number of allowed requests per period
+    , throttleSettingsPeriod = 1_000_000 -- 1 second
```
Comment (fosskers) [inline src/Chainweb/Chainweb.hs]: A simpler representation of the same value.

File: chainweb.cabal (hunk 1 above)
Comment (larskuhtz) [reply chainweb.cabal]: Is this still needed?

Comment (larskuhtz) [reply chainweb.cabal]: At least it should be only 1.0.3, I think.

**Comment by fosskers** [comment]
I'm going to rebase this.

File: src/Chainweb/Chainweb.hs [hunk 4]
```diff
@@ -611,6 +633,7 @@ withChainwebInternal conf logger peer rocksDb dbDir nodeid resetDb inner = do
                             , _chainwebThrottler = throttler
                             , _chainwebMiningThrottler = miningThrottler
                             , _chainwebPutPeerThrottler = putPeerThrottler
+                            , _chainwebLocalThrottler = localThrottler
```
Comment (larskuhtz) [inline src/Chainweb/Chainweb.hs]: Are we actually serving this middleware?

Comment (fosskers) [reply src/Chainweb/Chainweb.hs]: I'll doublecheck.

Comment (larskuhtz) [reply src/Chainweb/Chainweb.hs]: It must be added to this:

```Haskell
IrunChainweb cw = do
    logg Info "start chainweb node"
    concurrently_
        -- 1. Start serving Rest API
        (serve
            $ httpLog
            . _chainwebRouteBlacklist cw
            . throttle (_chainwebPutPeerThrottler cw)
            . throttle (_chainwebMiningThrottler cw)
            . throttle (_chainwebThrottler cw)
        )
```

**Comment by fosskers** [comment]
Currently incomplete.

**Pull request merged**